### PR TITLE
[WIP] Implement splittable PRNGKey

### PR DESCRIPTION
**Pull request by ejguan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62576 [WIP] vmap for randn with PRNGKey
* #62575 [WIP] Implement randn with PRNGKey
* **#62574 [WIP] Implement splittable PRNGKey**
* #62573 [WIP] Expose Philox as function to return uint64_t

- Add `SplittableRNGKey` dispatch key for `PRNGKey`
- Add an API of `torch.PRNGKey` to return a Tensor with the new `SplittableRNGKey` dispatch key
- Add `is_rng_key` to TensorBase to determine if the Tensor is a SplittableRNGKey or normal Tensor
- Add `split_key` function to split the old `PRNGKey` to a new `PRNGKey` by adding a new dimension with size of counter (using philox)
  - The reason that it's not feasible to use `split` overload is the duplicate signature:
    - `split.Tensor(Tensor(a) self, int split_size, int dim=0) -> Tensor(a)[]` returns an array of Tensor.
    - `split_key(Tensor self, int ctr=1) -> Tensor` requires to return only a tensor
    - Potentially add `split` method here to make `split` working for `PRNGKey` https://github.com/pytorch/pytorch/blob/cf1f59452b05322425f606afcf6114b4b448b356/torch/_tensor.py#L493-L507

### CODE REVIEW

**Comment by ezyang** [comment]
Is there a design doc for this patch series?

**Comment by ejguan** [comment]
@ezyang 
Yes, here is the [Proposal](https://fb.quip.com/eg3LA7bblmhZ). And, Richard and I had a series of offline discussions about the proposal.

File: aten/src/ATen/native/Splitting.cpp [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+Tensor split_key(const Tensor& self, int64_t ctr) {
+  TORCH_CHECK(self.is_rng_key(), "`split_key` only accepts PRNGKey created from `torch.PRNGKey`");
+  std::vector<Tensor> keys(ctr);
+  auto sizes = self.sizes();
+  for (const auto i : c10::irange(ctr)) {
+    auto tensor = detail::_philox(self, i);
+    keys[i] = std::move(tensor);
+  }
+  auto res = at::stack(keys, 0);
```
Comment (zou3519) [inline aten/src/ATen/native/Splitting.cpp:33]: This should be parallelizable. E.g. we could create a `_phillox(const Tensor& self, const Tensor& counters)` binary operator

File: aten/src/ATen/SplittingFallback.cpp [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    }
+    auto tensor = std::move(ivalue).toTensor();
+    tensor._set_rng_key(true);
+    (*stack)[stack_start + i] = std::move(tensor);
+  }
+}
+
+TORCH_LIBRARY_IMPL(_, SplittableRNGKey, m) {
+  m.fallback(torch::CppFunction::makeFromBoxedFunction<&splittable_rng_fallback>());
+}
```
Comment (zou3519) [inline aten/src/ATen/SplittingFallback.cpp:29]: This is cool! If we didn't want this to work for all PyTorch operations, then maybe we can have an allowlist of operations

Comment (ejguan) [reply aten/src/ATen/SplittingFallback.cpp:29]: Yeah, at least making view operations and `split_key` available. And the in-place op should be blocked.

**Pull request closed without merging**